lucene/core/src/java/org/apache/lucene/codecs/lucene80/IndexedDISI.java

```diff
@@ -0,0 +1,542 @@
+    flush(DocIdSetIterator.NO_MORE_DOCS >>> 16, buffer, 1, out);
+    // offset+index jump-table stored at the end
+    return flushBlockJumps(jumps, lastBlock, out, origo);
+  }
+
+  // Adds entries to the offset & index jump-table for blocks
+  private static long[] addJumps(long[] jumps, long offset, long index, int startBlock, int endBlock) {
+    jumps = ArrayUtil.grow(jumps, endBlock +1);
+    final long jump = (index << BLOCK_INDEX_SHIFT) | offset;
+    for (int b = startBlock; b < endBlock; b++) {
+      jumps[b] = jump;
+    }
+    return jumps;
+  }
+
+  // Flushes the offet & index jump-table for blocks. This should be the last data written to out
+  // This method returns the blockCount for the blocks reachable for the jump_table or -1 for no jump-table
+  private static short flushBlockJumps(long[] jumps, int blockCount, IndexOutput out, long origo) throws IOException {
+    if (blockCount == 1) { // A single jump is just wasted space so we ignore that
+      blockCount = 0;
+    }
+    for (int i = 0 ; i < blockCount ; i++) {
+      out.writeLong(jumps[i]);
+    }
+    // As there are at most 32k blocks, the count is a short
+    // The jumpTableOffset will be at lastPos - (blockCount * Long.BYTES)
+    return (short)blockCount;
+  }
+
+  /** The slice that stores the {@link DocIdSetIterator}. */
+  private final IndexInput slice;
+  private final int jumpTableEntryCount;
+  private final RandomAccessInput jumpTable; // Skip blocks of 64K bits
+  private final byte[] rankTable = new byte[256]; // rank-table (logically 128 * 16 bit) for DENSE blocks
+  private final long cost;
+
+  /**
+   * @param in backing data.
+   * @param offset starting offset for blocks in backing data.
+   * @param length the number of bytes in the backing data.
+   * @param jumpTableEntryCount the number of blocks convered by the jump-table.
+   * @param cost normally the number of logical docIDs.
+   */
+  IndexedDISI(IndexInput in, long offset, long length, int jumpTableEntryCount, long cost) throws IOException {
+    this(in.slice("docs", offset, length), jumpTableEntryCount, cost);
+  }
+
+  /**
+   * This constructor allows to pass the slice directly in case it helps reuse.
+   * see eg. Lucene80 norms producer's merge instance.
+   * @param slice backing data, starting from position 0 to slice.length
+   * @param jumpTableEntryCount the number of blocks convered by the jump-table.
+   * @param cost normally the number of logical docIDs.
+   */
+  IndexedDISI(IndexInput slice, int jumpTableEntryCount, long cost) throws IOException {
+    this.slice = slice;
+    this.cost = cost;
+
+    this.jumpTableEntryCount = jumpTableEntryCount;
+    //slice.seek(slice.length()-Short.BYTES);
+    if (jumpTableEntryCount <= 0) {
+      jumpTable = null;
+    } else {
+      long jumpTableOffset = slice.length()-jumpTableEntryCount*Long.BYTES;
+      jumpTable = slice.randomAccessSlice(jumpTableOffset, slice.length()-jumpTableOffset);
+    }
```

Comment:
you will need to support reuse for jumpTable as well. Doc values are special because they get pulled when writing postings in order to index impacts, and we don't want to create a new slice for every indexed term. I believe the test that tests slice creation only doesn't fail because we don't create jump tables if there aren't many docs.

Reply:
It worked before because jumpTables are relative to the length of the slice, not the current position. Anyway, now I understand what is happening. Re-use of jumpTables is in the next commit.